Add Android AVD / emulator actions

AliSoftware opened the pull request:
Adds a helper and multiple actions to help us install Android system-images, create AVDs dynamically, and boot and shutdown Android emulators.

These are used by our `screenshots` lane in https://github.com/wordpress-mobile/WordPress-Android/pull/17096 to automate capture of raw screenshots.

> EDIT: Reworked to improve many error cases (e.g. `sdkmanager` installed via `brew` but `emulator` installed by Android Studio, leading to multiple SDK roots; or giving hints for the `PANIC: Broken AVD` message; etc.) and to support dynamic emulator TCP ports (i.e. non-hardcoded `emulator-5554` serial) to support running multiple emulators in parallel 💪 

## To Test

### Base Scenario

 - Switch to the `tooling/screenshots` branch in WPAndroid (from https://github.com/wordpress-mobile/WordPress-Android/pull/17096)
 - Run `bundle install` to make sure you point the `fastlane/Pluginfile` and `release-toolkit` in that WPAndroid branch to this toolkit's `screenshots/emulator-helper` branch
 - Run `bundle exec fastlane screenshots app:wordpress locale:zh-CN` from the WPAndroid repo
 - Ensure that the screenshot automation finishes and worked (it can take a couple of minutes†…), in particular:
   - Check that the logs mention calling the `android_create_avd` action, installing the system image and creating the AVD successfully, for each case (phone, tablet)
   - Check that the logs mention calling the `android_launch_emulator` action, booting the emulator then waiting for it to finish booting, for each case (phone, tablet)
   - Check that the lane ended up creating the raw screenshots in Chinese for both phone and tablet device sizes successfully.
   - Once the full lane completes and the screenshot UI tests have finished running, check that the raw screenshots images have been generated and pulled from the emulator's sdcard, and are present in the HTML summary file.

> **Note**: _† For some reason it appears that sometimes there's a long pause between when the unit test has started and the app has launched… and when the first click of the UI test scenario finally starts. I'm still not sure what could explain those delays (they are not consistent and are hard to debug), so for the time just don't be surprised that it could sometime be between 2 to 8 minutes between when the app's first screen shows in the emulator and when the UI test finally starts its clickity-clackity work._ 🙃 

### Variants / configurations

Please consider testing all these different setups, to ensure all my error handling of possible scenarios encountered on different user's machines or CI hosts will be covered

 - 💻  Test on both an Intel and an M1 machine
    - especially to validate that the installation of the appropriate Android System Image by `sdkmanager` installs the appropriate system image depending on the architecture of the machine you run those lanes from.
 - 🔧  Test with different ways of installing the Android Command Line Tools
    - Uninstall the command line tools like `sdkmanager` and `avdmanager`, then re-run the `screnshots` lane and confirm that the action errors, with a suggestion to install them
    - Install the Android command line tools [using `brew`](https://formulae.brew.sh/cask/android-commandlinetools) — while you keep `emulator` installed by Android Studio and reachable from terminal (either via your `$PATH` or by having `$ANDROID_SDK_ROOT` defined). Then re-run the `screenshots` lane and confirm that the action errors, suggesting that you might have two different Android SDK locations (brew and AS) and that `sdkmanager` and `emulator` might not be installed in the same one. Then `brew uninstall` the Android command line tools once done.
    - Finally, re-install the Command Line Tools via the Android Studio's "SDK Manager" window, and test the lane again. That time it should succeed.
 - 🏎️  Test to run multiple emulators in parallel
    - Start any AVD from Android Studio that is neither `Pixel_3_API_28` nor `Nexus_9_API_28`. The goal here is to have the TCP port `5554` — which is usually the default one / first picked by `adb` to connect to an emulator — already in use.
    - In one terminal window, run `bundle exec fastlane screenshots app:wordpress locale:ar device:phone`
    - While the first command is still running, run `bundle exec fastlane screenshots app:wordpress locale:ar device:tenInch` in another terminal window
    - Verify that each command found the correct serial (e.g. `emulator-5556` for the phone one and `emulator-5558` for the tablet one — since `emulator-5554` was likely the one already used by the dummy AVD you launched from AS in step one)

AliSoftware committed:
Add Android AVD / emulator actions

AliSoftware committed:
Fix rubocop violation on comment

AliSoftware committed:
Add CHANGELOG entry

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,162 @@
+        # @return [String] emulator serial number corresponding to the launched AVD
+        #
+        def launch_avd(name:, cold_boot: true, wipe_data: true)
+          port = '5554'.freeze
+          serial = "emulator-#{port}"
+
+          shut_down_emulators!(serials: [serial]) # To ensure we can launch one on the port 5554 (as it's hardcoded for simplicity)
```
Known limitation of the current implementation: the port is hardcoded to `5554` — which is why we need to ensure we shut down existing emulators if any first — because it ended up not being that trivial to make it dynamic.

If I let `emulator` choose the ports by itself (instead of explicitly providing them via `-port` flag on line 77), which would ideally be a more flexible alternative, I then couldn't find an easy and consistent/safe enough way to then retrieve the port that was used. I considered looking at the device list before and after to compare and detect the new one, but:
 - There's a small risk that the new emulator disappears from `adb devices -l` for a brief moment while it's booting and setting itself up during booting sequence
 -  That's not a strong guarantee because the user might also have other emulators that are in the process of booting or shutting down in parallel, which would lead to some race condition on checking the device list to compare
 - I've also took a quick look at if there were some command that would allow me to get a list of AVD names <-> adb ports used, which maybe we could use to try and match the AVD name with its port.
    - I've found some ideas around looping on each emulator, and call `adb -s emulator-#{port} emu avd name` command on each to get the name of the AVD…
    - …but that supposes the emulator has already finished booting (and we can't use `adb -s emulator-#{port} wait-for-device` to wait for the AVD we just launched to finish being ready… because we need to know the serial and thus the port to call that one… so catch-22)
    - …and we could consider looping over all the emulators found in `adb devices -l`, and call `adb -s emulator-#{port}` on each… but if we happen to do that loop before the AVD we just launched hasn't finished booting and connecting to `adb`, that won't find it and we'd have to do the loop all over again (using our `retry_loop` helper, probably), and calling `adb emu avd name` all over again… doesn't feel very efficient nor stable?

So in the end, I'm open to any ideas on how to improve flexibility here, or ways to find out the serial of an AVD we just booted without risking falling into the cracks and race condition / timing traps highlighted above. (@ParaskP7  any ideas or `adb` magic commands you know could help?)

---

The impact of this known limitation of hard-coding the port for now is that we won't be able to run screenshot tests on the same machine in parallel (as those would try to boot multiple emulators). If we find a working solution to make it possible for it to be dynamic (and thus allow screenshot parallelism), we can always add the feature in a subsequent PR.

AliSoftware replied:
EDIT: I've ended up fixing that one with latest commits on this PR after all 🎉

AliSoftware commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,162 @@
+          # Get the intersection of the set of running emulators with the ones we want to shut down
+          emulators_list &= serials unless serials.nil?
+          emulators_list.each do |e|
+            Actions.sh(@tools.adb, '-s', e, 'emu', 'kill') { |_| } # ignore error if no emulator with specified serial is running
+
+            # NOTE: Alternative way of shutting down emulator would be to call the following command instead, which shuts down the emulator more gracefully:
+            # `adb -s #{e} shell reboot -p` # In case you're wondering, `-p` is for "power-off"
+            # But this alternate command:
+            #  - Requires that `-no-snapshot` was used on boot (to avoid being prompted to save current state on shutdown)
+            #  - Disconnects the emulator from `adb` (and thus disappear from `adb devices -l`) for a short amount of time,
+            #    before reconnecting to it but in an `offline` state, until `emulator` finally completely quits and it disappears
+            #    again (for good) from `adb devices --list`.
+            # This means that so if we used alternative, we couldn't really retry_loop until emulator disappears from `running_emulators` to detect
+            # that the shutdown was really complete, as we might as well accidentally detect the intermediate disconnect instead.
```
Another area where I'd love this to be cleaner.

Especially since some old versions of the `emulator` binary end up triggering a CrashReporter alert on macOS when we send the `emu kill` command to the emulator to shut it down that way… and having a big alert on your OS every time and on every iteration of the loop that boots-then-shutdown an emulator for each locale and device that raw screenshots are taken in is… not ideal

Luckily, that CrashReporter alert I've seen in the past seems to have disappeared and not happening anymore with more recent versions of `emulator` (I'm currently using `emulator 31.3.10.0 (build_id 8807927)`), so it's probably safe to use `emu kill` after all instead of trying to be more clever with `adb shell reboot -p` 🤞

AliSoftware committed:
Fix error handling when searching for Android tools

AliSoftware committed:
Always install system-image in current SDK path

This is especially to avoid issues with cases where:
 - You have `ANDROID_SDK_ROOT` set
 - But you used `brew install --cask android-commandlinetools` to install `sdkmanager` (instead of installing it via Android Studio), which installs `sdkmanager` in `$(brew --prefix)` and not in your `$ANDROID_SDK_ROOT`
 - As a result, `sdkmanager` would default to install the `system-image` in `$(brew --prefix)/share/android-commandlinetools` and not in your `$ANDROID_SDK_ROOT`, but when `emulator` will try to boot your AVD, it will search the system-image in `$ANDROID_SDK_ROOT` and fail to find it

This particular situation would lead to the `emulator` failing to start and print a "PANIC: Broken AVD system path" message, while `adb wait-for-device`, waiting for the boot to finish in parallel, would wait forever and never exit.

AliSoftware committed:
Fix `emulator` logs filtering

To still remove most of the noise, but still log ERROR and PANIC lines, especially to print the `PANIC: Broken AVD system path` and other boot errors like "cannot find kernel file" that ends up failing to boot the emulator.

If we don't print those. the user will see in their fastlane logs that the emulator command was called, then that the `adb wait-for-device` was called, but then adb will be waiting for that device forever since `emulator` never really launched.
In those cases it's thus critical to log the errors to help understand that it's no use to wait forever and to have hints of what to fix.

AliSoftware committed:
Use a `Thread` + `Open3.popen2` to launch emulator

So that we can process the stdout and stderr of the `emulator` process in Ruby, rather rhan relying on spawning a subshell and rely on piping to `sed` and having to use `system` and `&` for running in the background… which was not pretty.

Here by using `Thread.new` we're still able to launch the emulator in the background, but we're also able to process its `stdout` and `stderr` directly from our ruby code, by calling `Open3.popen2e`, reading the `stdout`+`stderr` as an `IO` line by line (`oe.readline`) as a line stream, and print emulator's log lines as they arrive to u to be able to filter then and add a prefix, all using Ruby code instead of `sed`

This also allows us to print the exit code of the `emulator` process if it didn't exit properly (non-zero status code).

As a final bonus, the `emulator` subprocess that has been opened by `Open3.popen2s` within that `Thread` will be automatically killed (being a child process) when the parent process (fastlane) terminates (successfully or not). Which was not the case when we used `system('emulator … &') to run the emulator as a background job.

AliSoftware committed:
Allow `launch_avd` to use TCP port auto-detect

If `port` is not provided explicitly, will let `emulator` pick the TCP port.
Then we will loop on each emulator device to find the one with our AVD name in order to find its serial to use in further `adb -s` commands

AliSoftware committed:
Improve error handling + add hints on panics

AliSoftware committed:
Restore sdkmanager trying to outsmart SDK path

It's not for us to try to work around broken installation that are a mix of Android Studio and `brew install` leading to multiple Android SDK roots on the system.

So it's better not to try to be smart by forcing `--sdk_path` on `sdkmanager` here, especially since the only impact that it will have if the user installed the android command-line-tools via `brew --cask` and not Android Studio, is that this will just push the can down the road and crash during the call to `avdmanager`…

AliSoftware committed:
Fix find_serial method

AliSoftware committed:
Better error handling

AliSoftware committed:
Fix `find_serial`

And improve logging around it

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,198 @@
+          end&.serial
+        end
+
+        # Trigger a shutdown for all running emulators, and wait until there is no more emulator running.
```
```suggestion
        # Trigger a shutdown for all running emulators, and wait until there is no more emulators running.
```

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,198 @@
+          @tools = Fastlane::Helper::Android::ToolsPathHelper.new
+        end
+
+        # Installs the system-image suitable for a given Android `api`, with `google_apis`, and for the current machine's architecture
```
Nitpick: To make this consistent with the usage [below](https://github.com/wordpress-mobile/release-toolkit/pull/409/files#diff-d271fcb6de300236feaec1c6b889596caf09c69201f3f8d89a692cd0496dd8ddR170)


```suggestion
        # Installs the system-image suitable for a given Android `api`, with Google APIs, and for the current machine's architecture
```

Personally, I prefer "Google APIs" over `google_apis` because the latter makes me think of a method parameter.

AliSoftware replied:
Right. I used this wording for `google_apis` here because that's how it's specified in the system-images qualifier expected to be passed to `sdkmanager --install <qualifier>`. But I can see the rationale for consistency with below and with differentiating `Google APIs` as a name from `api` as a parameter (and also the image qualifier could be seen as an implementation details… even if I thought the explicitness of the package component specifier here could help explain better which package this method installed)

```
sdkmanager --list_installed
…
system-images;android-28;google_apis;arm64-v8a | …
…

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,198 @@
+          UI.success('All emulators are now shut down.')
+        end
+
+        # Find the system-images package for the provided API, with Google APIs, and matching the current platform/architecture this lane is called from.
```
Nitpick, to make it consistent with the usage [above](https://github.com/wordpress-mobile/release-toolkit/pull/409/files#diff-d271fcb6de300236feaec1c6b889596caf09c69201f3f8d89a692cd0496dd8ddR17)

```suggestion
        # Find the system-images package for the provided `api`, with Google APIs, and matching the current platform/architecture this lane is called from.
```

Personally, I prefer using `api` over API because it makes it clearer that we are referring to the method paratmer.

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_emulator_helper.rb:
```diff
@@ -0,0 +1,198 @@
+            platform = `uname -m`.chomp
+            all_packages = `#{@tools.sdkmanager} --sdk_root=#{@tools.android_sdk_root} --list`
+            package = all_packages.match(/^ *(system-images;android-#{api};google_apis;#{platform}(-[^ ]*)?)/)&.captures&.first
+            UI.user_error!("Could not find system-image for API `#{api}` and your platform `#{platform}` in `sdkmanager --list`. Maybe Google removed it for download and it's time to update to a newer API?") if package.nil?
```
"Maybe ... it's time to update ..." is an appropriate suggestion, but sets the precedent for adding similar suggestions to other error messages.

Is there a particular reason why you added it? Would users get less value from the tool if it wasn't there?

AliSoftware replied:
Mostly because last time I god a `PANIC: Broken AVD system image` errors when working on screenshots automation and trying to use the system-images that were defined in our scripts, it took me a very long time to understand what was broken (ref: paqN3M-cM-p2#broken-avd-system-image), and get why it suddenly didn't work on my machine while it worked for other team members (who had the system image installed on their machine from them downloading it a long time ago before it got removed from Google's servers), before I finally understand that that the only way to fix that was to update the configuration.

Given Google apparently removes old system images from their download servers from time to time, this makes this screenshot automation lane able to work on machines that installed the system image from a previous download but fail on machines that try to download it for the first time, and make for that lane to work one day, but break if you try to re-run it later without any change, i.e. it depends on Google download servers changes that makes old successful builds (e.g. from CI) not always reproducible for that reason. That's why I thought this suggestion was important to mention in this case.

mokagio replied:
Fair 👌

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_tools_path_helper.rb:
```diff
@@ -0,0 +1,59 @@
+        attr_reader :android_sdk_root
+
+        def initialize(sdk_root: nil)
+          @android_sdk_root = sdk_root || ENV['ANDROID_HOME'] || ENV['ANDROID_SDK_ROOT'] || ENV['ANDROID_SDK']
```
Where does `ANDROID_SDK` come from? I couldn't find it in the [Environment variables](https://developer.android.com/studio/command-line/variables#envar) documentation.

Also, the docs say that `ANDROID_SDK_ROOT` is deprecated. Should we avoid looking for it?

> `ANDROID_HOME`: Sets the path to the SDK installation directory. Once set, the value does not typically change, and can be shared by multiple users on the same machine. `ANDROID_SDK_ROOT`, which also points to the SDK installation directory, is deprecated. If you continue to use it, Android Studio and the Android Gradle plugin will check that the old and new variables are consistent.

mokagio replied:
By the way, two points:

1. It goes without saying, but, as evidenced by the fact that I haven't requested changes, I don't think this is a big issue. Happy for a followup PR _if_ you consider the suggestion.
2. _If_ implementing the suggestion, would it be useful to rename the various `android_sdk_root` to `android_home` for consistency?

AliSoftware replied:
Good question, I don't remember where I found that from tbh, but I think I [borrowed that from `fastlane` itself](https://github.com/fastlane/fastlane/search?q=ANDROID_SDK)

> Also, the docs say that ANDROID_SDK_ROOT is deprecated.

Oh, funny, I always thought `ANDROID_HOME` was the one being deprecated and `ANDROID_SDK_ROOT` was the new hotness in town.

Given what you linked in the documentation but also the order those env vars are checked by `fastlane` (and thus by this line here which I borrowed from there), I guess I'll have to rename all my env vars in my machine, `~/.zshrc`, and personal scripts I have around on my Mac (and maybe in other scripts in this repo like [this one](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/tools/rename_apk_aab.sh#L33), and check which ones we use on our CI for our Android repos too)

> Should we avoid looking for it? 

We still definitively need to keep looking for it, because I think me and many Android devs still use `ANDROID_SDK_ROOT` in many places in their config on their machine.

AliSoftware replied:
Fixed in 7cba9a00eb16cfdba8193e5e664f313b79e2dd8e

Then applied a similar fix in 97d32333ad679942d9d8d594139cd8ed04da8cee on other actions that predates this PR (`android_send_app_size_metrics`) but had the same incorrect assumption (that `ANDROID_HOME` was the deprecated one, while `ANDROID_SDK_ROOT` is the deprecated one and `ANDROID_HOME` is the newest)

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_shutdown_emulator_action.rb:
```diff
@@ -0,0 +1,48 @@
+          FastlaneCore::ConfigItem.new(key: :serial,
+                                       env_name: 'FL_ANDROID_SHUTDOWN_EMULATOR_SERIAL',
+                                       description: 'The serial(s) of the emulators to shut down. If not provided (nil), will shut them all down',
+                                       type: Array,
```
Given the type is `Array`, how would you feel about renaming the parameter so `:serials`?

Or, is there a convention in Fastlane to use the singular form because it supports passing a single value and converts it to an array of one element under the hood?

AliSoftware replied:
Yeah I actually wondered about this. Ended up using singular because I figured in 95% of cases the call site will only provide either `nil` or a single `serial`, and providing multiple serials (but a subset of all of them) would be a rather rare case.

But I guess using `serials` make more sense for consistency with the param being an `Array` after all. Will update.

AliSoftware replied:
Fixed in 7c152d3705fa050b5f50196a60a7c8bc4000441e (then updated call site in https://github.com/wordpress-mobile/WordPress-Android/commit/b6346ab0eaf45891c8c66e47301b6529a93807b8

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/actions/android/android_create_avd_action.rb:
```diff
@@ -0,0 +1,90 @@
+                                       default_value: '512M'),
+          FastlaneCore::ConfigItem.new(key: :system_image,
+                                       env_name: 'FL_ANDROID_CREATE_AVD_SYSTEM_IMAGE',
+                                       description: 'The system image to use (as used/listed by `sdkmanager`). Defaults to the appropriate system image given the API level requested and the current machine\'s architecture',
```
Random question out of curiosity: Why did you go for an escaped `'` for the Saxon possessive as opposed to using double quotes in the strings?

AliSoftware replied:
No reason really. I think it's a matter of finding balance as to when to switch to double-quotes when it's only to avoid escaping apostrophes in the String. If it's only one to be escaped, it's probably fine to keep single-quotes to avoid needing string interpolation to kick in for nought, when it's more than one, it's easier to switch to double-quotes to avoid escaping each one? But I don't have a strict rule for it tbh.

Another alternative would be to use alternate delimiters that still don't do string interpolation — since Ruby supports multiple string delimiters, interpolated or not. For example `%(…)` instead of `'…'` is a good choice when it comes to not needing interpolation but still wanting to avoid escaping.
```suggestion
                                       description: %(The system image to use (as used/listed by `sdkmanager`). Defaults to the appropriate system image given the API level requested and the current machine's architecture),
```

mokagio reviewed (approved):
Leaving a tentative approval to unblock your work.

I've looked at the code and it looks good. 

I meant to test this by running some actions and I even started one, but I'm running out of time for the day.

<img width="1985" alt="image" src="https://user-images.githubusercontent.com/1218433/190961972-eb9b221c-d44c-4c48-b8cb-91493fff4b70.png">

mokagio commented:
As I submitted my review, I noticed my command had failed. The failure looks to me like an HTTP out of the control of the release toolkit. The error message it printed is worth sharing, because of it's stacktrace formatting:

![image](https://user-images.githubusercontent.com/1218433/190962282-a19338e8-c45d-4bc3-a32a-a9ff3397923a.png)

![image](https://user-images.githubusercontent.com/1218433/190962319-9cee4636-f021-4b6e-9e19-f29a94c407a8.png)

mokagio commented:
> The failure looks to me like an HTTP out of the control of the release toolkit

I retried the action and this time it succeeded, which is a point in favor of my conjecture about the nature of the previous failure:

![image](https://user-images.githubusercontent.com/1218433/190963013-94e2a3af-7cba-4379-816b-8d471c9d4e38.png)

AliSoftware committed:
Fix typo in comment

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware committed:
Improve some wording in comments

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

AliSoftware commented:
> Leaving a tentative approval to unblock your work.

🙇 

> I meant to test this by running some actions and I even started one, but I'm running out of time for the day.
> 
> <img alt="image" width="1985" src="https://user-images.githubusercontent.com/1218433/190961972-eb9b221c-d44c-4c48-b8cb-91493fff4b70.png">

@mokagio  Can you tell me what command you ran to do that test? I'm surprised that in your capture above we see `fastlane` asking you for values for the command. Did you e.g. create a custom `lane :test` in your `Fastfile` for testing this action and then ran it, which is why it asked? (as opposed to following the testing steps I mention and using the `screenshots` lane in my WPAndroid PR, which are not supposed to ask for the `device_model` and `api_level` values and are supposed to provide them explicitly during the call of this action)

AliSoftware commented:
> The error message it printed is worth sharing, because of it's stacktrace formatting:

Yeah that's a quirk of how `sdkmanager` includes `\r` (without `'n`) in the `stdout` to do this ASCII progress bar, and fail to introduce a `\n` in `stdout` before printing its error message when it fails. And then `fastlane`'s `UI.shell_error!` prints the last "line" (as delimited by `\n`, rightfully) where the error appeared, hence including those extra `\r` that were not separated by a `\n` from the actual error message. I guess we could try and fix fastlane's `UI.shell_error!` to split around `\r` too though technically that would be a bit incorrect to do that for all cases… and we won't be able to fix the issue at its core (i.e. in `sdkmanager` itself), so… I guess the error in those cases will have to include those extra `[ASCII progress bar ===]\r` pseudo-lines in its output 🙃

mokagio commented:
@AliSoftware 

> @mokagio Can you tell me what command you ran to do that test?

```
bundle exec fastlane run android_create_avd
```

> I'm surprised that in your capture above we see fastlane asking you for values for the command.

Given I called the action with no parameters, that should be the expected behavior, right?

AliSoftware commented:
> @AliSoftware
> 
> > @mokagio Can you tell me what command you ran to do that test?
> 
> ```
> bundle exec fastlane run android_create_avd
> ```
> 
> > I'm surprised that in your capture above we see fastlane asking you for values for the command.
> 
> Given I called the action with no parameters, that should be the expected behavior, right?

Ah right, if that's how you called it, then yes, totally 👍

AliSoftware committed:
Remove use of `--sdk_root` in `sdkmanager` call

As discussed in previous commit, we should now assume anything about the user's machine setup, and should let the default path and SDK resolution mechanism from the tools do their job. If the user has a broken setup (like multiple SDK roots / ANDROID_HOME for different tools), it's the user's responsibility to fix their setup, not for us to try to fix it (and make things even more obscure if our fix doesn't do as intended)

AliSoftware committed:
Rename `android_sdk_root` to `android_home`

As suggested in https://github.com/wordpress-mobile/release-toolkit/pull/409#discussion_r973879702

AliSoftware committed:
DRY some methods in `Android::ToolsPathHelper`

AliSoftware committed:
Make other action use `Android::ToolsPathHelper`

AliSoftware committed:
Use plural form for `Array`-typed action option

As suggested in https://github.com/wordpress-mobile/release-toolkit/pull/409#discussion_r973878479

AliSoftware committed:
Fix tools detection in $PATH

AliSoftware commented:
Merging this without waiting for someone else with M1 to test after all, to be able to move forward with this project.
